File: lib/android/src/main/java/com/wix/crypto/utilities/DerivationUtils.java
package com.wix.crypto.utilities;

import com.wix.crypto.Constants;
import com.wix.crypto.Crypto;
import com.wix.crypto.custom.Pair;

import java.util.Arrays;

/**
 * Created by hagai on 11/05/2020.
 */
public class DerivationUtils {

    public static final byte[] getKeyId(byte[] masterKey) {

        byte[] message = Constants.ENCODED_STRINGS.get("id");
        byte[] result = Crypto.HMAC(masterKey, message);

        return Arrays.copyOf(result, Constants.KEY_LEN);

    }

    public static final byte[] getMasterKeyCommitment(byte[] keyId, byte[] userId) {

        byte[] message = BytesUtils.byteConcatination(userId, Constants.ENCODED_STRINGS.get("idCom"));
        byte[ ] keyMasterCommitment = Crypto.HMAC(keyId, message);

        return Arrays.copyOf(keyMasterCommitment, Constants.KEY_LEN);
    }

    public static byte[] getKeyMasterVerification(byte[] masterKey) {

        byte[] message = Constants.ENCODED_STRINGS.get("verifkey");
        byte[] keyMasterVerification = Crypto.HMAC(masterKey, message);

        return Arrays.copyOf(keyMasterVerification, Constants.KEY_LEN);
    }

    public static byte[] getNextDayMasterKey(byte[] masterKey, boolean isInstallDay) {


        byte[] message;
        if(isInstallDay){
            message = Constants.ENCODED_STRINGS.get("master0");
        }
        else {
            message = Constants.ENCODED_STRINGS.get("master");
        }
        byte[] dayMasterKey = Crypto.HMAC(masterKey, message);


        return Arrays.copyOf(dayMasterKey, Constants.KEY_LEN);
    }

    public static byte[] getKeyEpoch(byte[] preKey, byte[] commitKey, byte[] day, byte[] epoch) {

        byte[] messagePart1 = BytesUtils.byteConcatination(commitKey, day);
        byte[] messagePart2 = BytesUtils.byteConcatination(epoch, Constants.ENCODED_STRINGS.get("depoch"));
        byte[] message = BytesUtils.byteConcatination(messagePart1, messagePart2);

        byte[] epochKey = Crypto.HMAC(preKey, message);

        return Arrays.copyOf(epochKey, Constants.KEY_LEN);
    }

    public static byte[] getKeyCommitmentForDayI(byte[] keyMasterCommitment, byte[] day) {
        assert day.length == 4;

        byte[] zeroBytes = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
        byte[] message = BytesUtils.byteConcatination(day, zeroBytes);

        return Crypto.AES(keyMasterCommitment, message);
    }

    public static byte[] getKeyVerificationForDayI(byte[] keyMasterVerification, int dayI) {

        byte[] day = BytesUtils.numToBytes(dayI, 4);
        byte[] message = BytesUtils.byteConcatination(day, Constants.ENCODED_STRINGS.get("dverif"));

        byte[] result = Crypto.HMAC(keyMasterVerification, message);

        return Arrays.copyOf(result, Constants.KEY_LEN);
    }

    //TODO:: problem with Pair data type.
    public static Pair<byte[],byte[]> getEpochKeys(byte[] epochKey, int day, int epoch) {

        byte[] prefix = BytesUtils.byteConcatination(BytesUtils.numToBytes(day, 4), BytesUtils.numToBytes(epoch, 1));
        byte[] zeroByteEleven = new byte[] {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
        byte[] zeroByteTen = new byte[]    {0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};


        byte[] epochEnc = Crypto.AES(epochKey, BytesUtils.byteConcatination(prefix, zeroByteEleven));

        byte[] message = BytesUtils.byteConcatination(prefix, zeroByteTen);
        byte[] epochMac = Crypto.AES(epochEnc, message);

        return new Pair<byte[], byte[]>(epochEnc, epochMac);
    }

    public static byte[] getKeyForDayI(byte[] dayMasterKey) {

        byte[] dayKey = Crypto.HMAC(dayMasterKey, Constants.ENCODED_STRINGS.get("ddaykey"));
        return Arrays.copyOf(dayKey, Constants.KEY_LEN);
    }
}

File: lib/ios/Crypto/DerivationUtils.swift
//
//  DerivationUtils.swift
//
//  Created by Yonatan Rimon on 10/05/2020.
//  Copyright © 2020 iGates. All rights reserved.
//

import Foundation

class DerivationUtils
{

    class func get_key_master_com(key_id: [UInt8], user_id: [UInt8]) -> [UInt8]
    {
        return Crypto.hmac_sha256_firstItems(key: key_id, data: user_id + STRINGS["id_com"]!, numberOfItems: const_KEY_LEN);
    }


    class func get_key_epoch(pre_key: [UInt8], commit: [UInt8], day: [UInt8], epoch: [UInt8]) -> [UInt8]
    {
        return Crypto.hmac_sha256_firstItems(key: pre_key, data: commit + day + epoch + STRINGS["depoch"]!, numberOfItems: const_KEY_LEN);
    }


    class func get_key_commit_i(key_master_com: [UInt8], day: [UInt8]) -> [UInt8]
    {
//        return Crypto.encrypt(key_master_com, day + b'\x00' * 12)
        //day must be 4 bytes
        return Crypto.encrypt(key: key_master_com, plain: BytesUtils.pad(array: day, size: const_KEY_LEN));
    }


    class func get_epoch_keys(epoch_key: [UInt8], day: Int, epoch: Int) -> ([UInt8], [UInt8])
    {
        var prefix = BytesUtils.numToBytes(num: day, numBytes: 4) + BytesUtils.numToBytes(num: epoch, numBytes: 1);
        let epoch_enc = Crypto.encrypt(key: epoch_key, plain: BytesUtils.pad(array: prefix, size: const_KEY_LEN));
        //Crypto.encrypt(epoch_key, prefix + b'\x00'*11)
        prefix.append(UInt8(0x01 ));
        let epoch_mac = Crypto.encrypt(key: epoch_key, plain: BytesUtils.pad(array: prefix, size: const_KEY_LEN));
        //var epoch_mac = Crypto.encrypt(epoch_key, prefix + b'\x01' + b'\x00'*10)
        return (epoch_enc, epoch_mac);
    }


    class func get_key_i_verification(key_master_verification: [UInt8], day: Int) -> [UInt8]
    {
        return Crypto.hmac_sha256_firstItems(key: key_master_verification, data: BytesUtils.numToBytes(num: day, numBytes: 4) + STRINGS["dverif"]!, numberOfItems: const_KEY_LEN);
    }


    class func get_next_day_master_key(prev_master_key: [UInt8], install_day: Bool = false) -> [UInt8]
    {
        if (install_day)
        {
            return Crypto.hmac_sha256_firstItems(key: prev_master_key, data: STRINGS["master0"]!, numberOfItems: const_KEY_LEN);
        }
        else
        {
            return Crypto.hmac_sha256_firstItems(key: prev_master_key, data: STRINGS["master"]!, numberOfItems: const_KEY_LEN);
        }
    }

    
}
